Feature: Add configurable amount representations for Joda Money module

apankowski opened the pull request:
This PR adds the possibility to configure desired representation of (de)serialized `Money` instances to the Joda Money Jackson module.

It allows selection of one of 3 representations via the `JodaMoneyModule.withAmountRepresentation(AmountRepresentation)` method:

* `DECIMAL_NUMBER` - the default; amounts are (de)serialized as decimal numbers equal to the monetary amount, e.g. `12.34` for EUR 12.34,
* `DECIMAL_STRING` - amounts are (de)serialized as strings containing decimal number equal to the monetary amount, e.g. `"12.34"` for EUR 12.34,
* `MINOR_CURRENCY_UNIT` - amounts are (de)serialized as long integers equal to the monetary amount expressed in minor currency unit, e.g. `1234` for EUR 12.34, `12345` for KWD 12.345 or `12` for JPY 12.

The motivation for having such feature is that:

 1. depending on the client implementation, passing monetary amounts as decimal numbers poses a risk of involving floating-point number representation, which may lead to loss of precision & rounding errors, especially in cases when some form of manipulation or arithmetic on the amounts is necessary,
 1. some APIs as well as quite popular frontend monetary amount handling library [Dinero.js](https://dinerojs.com/) require providing monetary amounts in minor currency units,
 1. there are numerous APIs that use the decimal string (quoted decimal numbers) representation.

Noteworthy points:

 1. I tried to align with the existing code style, but may have missed something.
 1. I took the liberty of using parametrized tests with help of [JUnitParams](https://github.com/Pragmatists/JUnitParams/).
 1. I tried to be careful to design the public parts of the new code to not expose too much and allow extensibility, but it would be good to review it thoroughly.
 1. I believe the changes don't change the existing API in any breaking way nor do they change the existing behavior.
 1. I have branched off of `2.13` (since it was the default) and made changes to code on that branch - not sure if that's correct.
 1. I will leave comments for discussion wherever I'm not sure of something.
 1. I'm not sure about the minimal JDK version we're targeting. I assumed JDK 8+ and therefore used `Objects.requireNonNull` and lambdas.

I will be helpful for pointers, suggestions and remarks and will try to align with all of them :slightly_smiling_face:

cowtowncoder commented on joda-money/src/test/java/com/fasterxml/jackson/datatype/jodamoney/ModuleTestBase.java:
```diff
@@ -30,28 +26,14 @@
                 .addModule(new JodaMoneyModule());
     }
 
-    protected static MapperBuilder<?,?> jodaMapperBuilder(DateFormat df) {
```
Good catch :)

cowtowncoder commented on joda-money/src/main/java/com/fasterxml/jackson/datatype/jodamoney/AmountRepresenter.java:
```diff
@@ -0,0 +1,13 @@
+
+public interface AmountRepresenter<T> {
+
+    T write(Money money);
```
Either name is sub-optimal (nothing gets written), or (more likely I think) this should do actual writing via `JsonGenerator` directly. If former, I think this is something like `AmountConverter`, and methods could be something like `fromMoney()` and `toMoney()`.

cowtowncoder commented on joda-money/src/main/java/com/fasterxml/jackson/datatype/jodamoney/MoneySerializer.java:
```diff
@@ -50,11 +52,7 @@ private final void _writeFields(final Money money,
-        g.writeNumberField("amount", decimal.setScale(scale, RoundingMode.UNNECESSARY));
-        g.writeFieldName("currency");
-        context.defaultSerializeValue(money.getCurrencyUnit(), g);
+        g.writeObjectField("amount", amountRepresenter.write(money));
```
But doesn't this always chnge value from JSON Number into JSON String? And be essentially backwards-incompatible change?

This is why I think instead of `write()` conversion method it should be method that does write.
So, first `g.writeFieldName("amount")`, then value write by helper object.

cowtowncoder replied:
Ah. I did miss the paramaterization, yes, `Long` and `BigDecimal` will then be written using configured serializer which is fine.

cowtowncoder commented:
First of all: thank you for contributing this patch! I haven't as much time to spend on things as I'd like, but I hope to provide some feedback.

I think the goals make sense, and I appreciate care taken. I did have some concerns (which may be due to my misreading?), added comments.

One practical thing is that due to changes to API, this would need to go in 2.14. I haven't yet created a branch for it, though, so for time being `2.13` makes most sense (`master` is for Jackson 3.0 that is further out).

Another thing is that more commonly formatting aspects would be handled via `@JsonFormat` annotation (esp. its `shape`), but I realize it may not be extensible enough here. It would allow differentiating between Number and String representation. It _might_ work, since we do have:

1. `Shape.STRING`
2. `Shape.NUMBER_FLOAT`
3. `Shape.NUMBER_INT` (for "minor")
4. `Shape.NUMBER` would be ambiguous; I assume that'd be same as (2)

The main benefits would be that everything (?) would be configurable by type and/or property; you can annotate properties explicitly, or indicate default for given type.

So this would be something to consider if you have time. I can try to help fill in blanks on usage; there isn't much tutorial on how to do these unfortunately.

apankowski reviewed (commented):
thank you for the pointers, I've answered the comments and pushed an update :slightly_smiling_face: 

regarding:

> more commonly formatting aspects would be handled via @JsonFormat annotation (esp. its shape), but I realize it may not be extensible enough here. It would allow differentiating between Number and String representation. It might work, since we do have:
>
> * Shape.STRING
> * Shape.NUMBER_FLOAT
> * Shape.NUMBER_INT (for "minor")
> * Shape.NUMBER would be ambiguous; I assume that'd be same as (2)
>
> The main benefits would be that everything (?) would be configurable by type and/or property; you can annotate properties explicitly, or indicate default for given type.

I actually wanted to approach something similar in a follow-up PR (to keep them smaller), but I can do it in this one if you prefer.

what I wasn't sure was if it will be natural for users (or: won't they be surprised) that a `Money` field is always serialized as JSON object, but `@JsonFormat(shape = )` on such field controls not serialization of the field itself, but just the "amount" field inside of it. (I guess we could just document it of course.)

second thing I was considering was a dedicated annotation, something like:

```
@JsonMoney(amountRepresentation = MINOR_CURRENCY_UNIT)
Money grandTotal;
```

but I'm not sure if reading such custom annotation is feasible in (de)serializers, won't break something (e.g. caching of (de)serializers), etc.

what do you think is the way to go in Jackson's world? :slightly_smiling_face:

cowtowncoder commented:
Yeah sorry have had bit of burnout wrt OSS work so feedback has been slow. But I do want to help here as this looks like a good improvement.

So: on `@JsonFormat` yeah... maybe that'd be unexpected; there is no precedent here.
I think going with the initial idea is fine; even if annotation support was added it could be incremental feature and not replacement (so module can define global defaults).

On getting this merged: with 2.13.0 this needs to go in 2.14; I haven't yet created 2.14 branches but hope to do that soon.

I'll add this on my TODO list and try to remember update once I get 2.14 branch set up for `jackson-databind` and this module.

cowtowncoder commented:
Ok still being slow -- this is on top of my review list, hoping to get it merged this week.

cowtowncoder commented:
@apankowski Apologies for yet more delays, but I'm back. I think things are good; only had 2 suggestions / questions:

1. Would it make sense to allow specific "withAmountConverter()" for fully custom translations; in addition to more convenient enum-alternative?
2. Javadocs would be good for couple of classes; I'll add a brief comment.

These are fully optional; just LMK what you think and I can merge things as-is or if you want to add/change something let me know when things are ready.

cowtowncoder commented on joda-money/README.md:
```diff
@@ -44,3 +44,13 @@ Money amount = mapper.readValue("{\"currency\":\"EUR\",\"amount\":19.99}", Money
+
+* `DECIMAL_NUMBER` - the default; amounts are (de)serialized as decimal numbers equal to the monetary amount, e.g. `12.34` for EUR 12.34,
+* `DECIMAL_STRING` - amounts are (de)serialized as strings containing decimal number equal to the monetary amount, e.g. `"12.34"` for EUR 12.34,
+* `MINOR_CURRENCY_UNIT` - amounts are (de)serialized as long integers equal to the monetary amount expressed in minor currency unit, e.g. `1234` for EUR 12.34, `12345` for KWD 12.345 or `12` for JPY 12.
```
Maybe just also add specific code example with module registration on mapper.

cowtowncoder commented on joda-money/src/main/java/com/fasterxml/jackson/datatype/jodamoney/AmountConverter.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.math.BigDecimal;
+
+public interface AmountConverter<T> {
```
Javadocs here would be nice  (although mostly if this is offered as an extension point as I suggest -- if not, just internal type)

cowtowncoder commented on joda-money/src/main/java/com/fasterxml/jackson/datatype/jodamoney/AmountRepresentation.java:
```diff
@@ -0,0 +1,8 @@
+
+public enum AmountRepresentation {
+
+    DECIMAL_NUMBER,
```
Javadocs really needed I think; nothing fancy.

cowtowncoder commented on joda-money/src/main/java/com/fasterxml/jackson/datatype/jodamoney/DecimalNumberAmountConverter.java:
```diff
@@ -0,0 +1,32 @@
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+
+public final class DecimalNumberAmountConverter implements AmountConverter<BigDecimal> {
```
Brief Javadoc would be good.

cowtowncoder commented on joda-money/src/main/java/com/fasterxml/jackson/datatype/jodamoney/DecimalStringAmountConverter.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.math.BigDecimal;
+
+public final class DecimalStringAmountConverter implements AmountConverter<String> {
```
Brief javadoc.

cowtowncoder commented on joda-money/src/main/java/com/fasterxml/jackson/datatype/jodamoney/MinorCurrencyUnitAmountConverter.java:
```diff
@@ -0,0 +1,29 @@
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+
+public final class MinorCurrencyUnitAmountConverter implements AmountConverter<Long> {
```
Brief javadoc.

cowtowncoder commented on joda-money/src/main/java/com/fasterxml/jackson/datatype/jodamoney/MoneyDeserializer.java:
```diff
@@ -16,15 +16,19 @@
+    private final AmountConverter<?> amountConverter;
 
-    public MoneyDeserializer() {
+    public MoneyDeserializer(final AmountConverter<?> amountConverter) {
```
For backwards-compatibility may want to add `@Deprecated` no-args constructor (to be removed from 3.0)

apankowski replied:
thanks, good catch :+1: the class if public after all, I missed that somehow

cowtowncoder commented on joda-money/src/main/java/com/fasterxml/jackson/datatype/jodamoney/MoneySerializer.java:
```diff
@@ -1,23 +1,25 @@
     private static final long serialVersionUID = 1L;
 
-    public MoneySerializer() {
+    private final AmountConverter<?> amountConverter;
```
As with deserializer, for 2.x backwards compatibility would be good to keep deprecated no-args constructor (to be removed from 3.0).

apankowski replied:
ditto

cowtowncoder commented on joda-money/src/main/java/com/fasterxml/jackson/datatype/jodamoney/AmountConverter.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.math.BigDecimal;
+
+public interface AmountConverter<T> {
```
Just realized that type reference here is not very valuable since the common base type is `java.lang.Object`.
May be left but wondering if it'd make sense to just drop it.

apankowski replied:
my intention was to make the "output" type somehow prominent, but it's true that doesn't bring much value - will remove

apankowski commented:
@cowtowncoder no worries - thank you for the feedback :slightly_smiling_face: 

Will apply all agreed changes and let you know soon. I could use some guidance with visibility of the `AmountConverter`s:

1. leave as is: `AmountConverter` and implementations are public (we document them a bit since they become part of the contract)
2. leave (documented) `AmountConverter` interface public but keep specific implementations package-private
3. make `AmountConverter` and specific implementations package-private

I can't decide how much we'd like to make public at first and could use your experience here

cowtowncoder commented:
Good questions! Since we do not really expose any way to use them except via enums, my slight preference would be for (3). That leaves us freedom to change things more should there be need for more customizations, in ways that we haven't anticipated.

apankowski commented:
@cowtowncoder PR updated and ready for next round of feedback :slightly_smiling_face: 

Summary of changes:

* Added module registration code sample in REAMDE.
* Removed generics from `AmountConverter`.
* Reduced visibility of the amount representation-related classes to package-private _except_ `AmountRepresentation` enum and corresponding `JodaMoneyModule.withAmountRepresentation()` representation selection method.
* Brought back default public constructors for `Money(De)Serializer`s to maintain compatibility with existing public API. Didn't add the `@Deprecated` annotation on them since API clients won't have another option to construct the (de)serializers directly (if they do it with existing API). The only other (indirect) way is by registering the `JodaMoneyModule`. Do you think we should add `@Deprecated` on default constructors of the (de)serializers and point clients to registering the `JodaMoneyModule` instead?
* Added simple (but hopefully readable) JavaDocs to amount representation-related classes.

cowtowncoder commented:
Looks good! I'll try to add some `@since` annotations but that can be done post-merge.

cowtowncoder commented:
Thank you again @apankowski !
